File: 2023/java/senior/Senior4.java
import java.util.*;

public class Senior4 {
    private static List<Integer> parent = new ArrayList<>();
    private static List<Map<Integer, Integer>> matrix = new ArrayList<>();

    private static int find(int u) {
        if (u != parent.get(u)) {
            parent.set(u, find(parent.get(u)));
        }
        return parent.get(u);
    }

    private static long dijkstra(int n, int s, int e) {
        long[] dist = new long[n + 1];
        Arrays.fill(dist, -1L);
        Set<Integer> visited = new HashSet<>();
        List<Integer> paths = new ArrayList<>();

        dist[s] = 0;
        paths.add(s);

        while (!paths.isEmpty()) {
            long min = Long.MAX_VALUE;
            int minPath = -1;
            for (int path : paths) {
                if (min > dist[path]) {
                    min = dist[path];
                    minPath = path;
                }
            }
            paths.remove((Integer) minPath);

            if (visited.contains(minPath)) {
                continue;
            }
            visited.add(minPath);

            for (Map.Entry<Integer, Integer> i : matrix.get(minPath).entrySet()) {
                if (!visited.contains(i.getKey()) && (dist[i.getKey()] == -1 || dist[minPath] + i.getValue() < dist[i.getKey()])) {
                    dist[i.getKey()] = dist[minPath] + i.getValue();
                    paths.add(i.getKey());
                }
            }
        }

        return dist[e];
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int m = in.nextInt();
        List<Node> roads = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            Node road = new Node();
            road.u = in.nextInt();
            road.v = in.nextInt();
            road.d = in.nextInt();
            road.c = in.nextInt();
            roads.add(road);
        }

        roads.sort((Node a, Node b) -> {
            if (a.d != b.d) {
                return a.d - b.d;
            }
            return a.c - b.c;
        });

        long total = 0;
        for (int i = 0; i <= n; i++) {
            parent.add(i);
            matrix.add(new HashMap<>());
        }

        for (Node r : roads) {
            int u = r.u;
            int v = r.v;

            int set_u = find(u);
            int set_v = find(v);

            if (set_u != set_v) {
                parent.set(set_u, set_v);
                total += r.c;
                matrix.get(u).put(v, r.d);
                matrix.get(v).put(u, r.d);
            } else if (r.d < dijkstra(n, u, v)) {
                total += r.c;
                matrix.get(u).put(v, r.d);
                matrix.get(v).put(u, r.d);
            }
        }

        System.out.println(total);
    }

    private static class Node {
        int u, v, d, c;
    }
}

File: 2023/cpp/senior/senior4.cpp
#include <bits/stdc++.h>
using namespace std;

struct EDGE {
    int x, y, d, c;
};

vector<int> parent{};
vector<EDGE> roads{};
vector<unordered_set<int>> groups{};
vector<vector<pair<int, int>>> matrix{};

int find(int u) {
    if (u != parent[u]) {
        parent[u] = find(parent[u]);
    }
    return parent[u];
}

long dijkstra(int n, int s, int e) {
    vector<long> dist(n + 1, -1);
    unordered_set<int> visited{};
    priority_queue<pair<long, int>, vector<pair<long, int>>> pq;

    dist[s] = 0;
    pq.emplace(0, s);

    while (!pq.empty()) {
        int u = pq.top().second;
        pq.pop();

        if (visited.find(u) != visited.end()) {
            continue;
        }
        visited.insert(u);

        for (auto i : matrix[u]) {
            if (visited.find(i.first) == visited.end() && (dist[i.first] == -1 || dist[u] + i.second < dist[i.first])) {
                dist[i.first] = dist[u] + i.second;
                pq.emplace(-1 * dist[i.first], i.first);
            }
        }
    }

    return dist[e];
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    for (int i = 0; i < m; i++) {
        int u, v, d, c;
        cin >> u >> v >> d >> c;
        roads.emplace_back(EDGE{u, v, d, c});
    }

    sort(roads.begin(), roads.end(), [](EDGE a, EDGE b) -> bool {
        if (a.d != b.d) {
            return a.d < b.d;
        }
        return a.c < b.c;
    });

    long total = 0;
    for (int i = 0; i <= n; i++) {
        parent.emplace_back(i);
        matrix.emplace_back();
    }

    for (auto r : roads) {
        int u = r.x;
        int v = r.y;

        int set_u = find(u);
        int set_v = find(v);

        if (set_u != set_v) {
            parent[set_u] = set_v;
            total += r.c;
            matrix[u].emplace_back(v, r.d);
            matrix[v].emplace_back(u, r.d);
        } else if (r.d < dijkstra(n, u, v)) {
            total += r.c;
            matrix[u].emplace_back(v, r.d);
            matrix[v].emplace_back(u, r.d);
        }
    }

    cout << total << endl;
}